File: Python/DesignPattern/CreationalDesignPattern/FactoryDesignPattern.py
# Factory Design Pattern
from abc import ABC, abstractmethod


class INotification(ABC):
    @abstractmethod
    def send(self): pass


class SMSNotification(INotification):
    def send(self): print("Sending SMS notification")


class WhatsAppNotification(INotification):
    def send(self): print("Sending WhatsApp notification")


class SlackNotification(INotification):
    def send(self): print("Sending Slack notification")


class NotificationFactory:
    _registry: dict = {}

    @staticmethod
    def get_instance(type: str) -> INotification:
        if type == "SMS":
            NotificationFactory._registry.setdefault(type, SMSNotification())
        elif type == "whatsapp":
            NotificationFactory._registry.setdefault(type, WhatsAppNotification())
        elif type == "Slack":
            NotificationFactory._registry.setdefault(type, SlackNotification())
        return NotificationFactory._registry.get(type)


class NotificationService:
    def send_notification(self, type: str, message: str):
        notification = NotificationFactory.get_instance(type)
        if notification:
            notification.send()


if __name__ == "__main__":
    service = NotificationService()
    service.send_notification("SMS", "Hello!")
    service.send_notification("whatsapp", "Hello!")
    service.send_notification("Slack", "Hello!")

File: C++/DesignPattern/CreationalDesignPattern/FactoryDesignPattern.cpp
// Factory Design Pattern
#include <iostream>
#include <string>
#include <unordered_map>
#include <memory>
using namespace std;

class INotification {
public:
    virtual void send() = 0;
    virtual ~INotification() {}
};

class SMSNotification : public INotification {
public:
    void send() override { cout << "Sending SMS notification\n"; }
};

class WhatsAppNotification : public INotification {
public:
    void send() override { cout << "Sending WhatsApp notification\n"; }
};

class SlackNotification : public INotification {
public:
    void send() override { cout << "Sending Slack notification\n"; }
};

class NotificationFactory {
    static unordered_map<string, shared_ptr<INotification>> registry;
public:
    static shared_ptr<INotification> getInstance(const string& type) {
        if (type == "SMS" && !registry.count(type))
            registry[type] = make_shared<SMSNotification>();
        else if (type == "whatsapp" && !registry.count(type))
            registry[type] = make_shared<WhatsAppNotification>();
        else if (type == "Slack" && !registry.count(type))
            registry[type] = make_shared<SlackNotification>();

        return registry.count(type) ? registry[type] : nullptr;
    }
};

unordered_map<string, shared_ptr<INotification>> NotificationFactory::registry;

class NotificationService {
public:
    void sendNotification(const string& type, const string& message) {
        auto notification = NotificationFactory::getInstance(type);
        if (notification) notification->send();
    }
};

int main() {
    NotificationService service;
    service.sendNotification("SMS", "Hello!");
    service.sendNotification("whatsapp", "Hello!");
    service.sendNotification("Slack", "Hello!");
    return 0;
}
